File: backend/src/main/scala/com/skewwhiffy/auraltester/notes/NoteLength.scala
package com.skewwhiffy.auraltester.notes

import com.skewwhiffy.auraltester.fractions.Fraction
import scala.util.chaining._

object NoteLength {
  lazy val breve: NoteLength = NoteLength(2)
  lazy val semibreve: NoteLength = NoteLength(1)
  lazy val minim: NoteLength = NoteLength(Fraction(1, 2))
  lazy val crotchet: NoteLength = NoteLength(Fraction(1, 4))
  lazy val quaver: NoteLength = NoteLength(Fraction(1, 8))
  lazy val semiquaver: NoteLength = NoteLength(Fraction(1, 16))
  //noinspection SpellCheckingInspection
  lazy val demisemiquaver: NoteLength = NoteLength(Fraction(1, 32))
  //noinspection SpellCheckingInspection
  lazy val hemidemisemiquaver: NoteLength = NoteLength(Fraction(1, 64))

  def apply(length: Int): NoteLength = NoteLength(Fraction(length, 1))

  def apply(length: Fraction): NoteLength = new NoteLength(length)
}

class NoteLength(private val length: Fraction) {
  lazy val dotted: NoteLength = Fraction(3, 2)
    .pipe(it => it * length)
    .pipe(it => NoteLength(it))

  lazy val abc: String = length.topHeavyString

  override def equals(obj: Any): Boolean = obj match {
    case it: NoteLength => it.length == length
    case _ => false
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/model/note/NoteLength.kt
package com.skewwhiffy.notation.model.note

import com.skewwhiffy.fraction.Fraction

data class NoteLength(private val length: Fraction) {
  @Suppress("unused", "SpellCheckingInspection")
  companion object {
    val breve: NoteLength = NoteLength(2)
    val semibreve: NoteLength = NoteLength(1)
    val minim: NoteLength = NoteLength(Fraction(1, 2))
    val crotchet: NoteLength = NoteLength(Fraction(1, 4))
    val quaver: NoteLength = NoteLength(Fraction(1, 8))
    val semiquaver: NoteLength = NoteLength(Fraction(1, 16))
    val demisemiquaver: NoteLength = NoteLength(Fraction(1, 32))
    val hemidemisemiquaver: NoteLength = NoteLength(Fraction(1, 64))
  }

  constructor(length: Int) : this(Fraction(length, 1))

  val dotted
    get() = NoteLength((Fraction(3, 2) * length))

  val abc: String = length.topHeavyString
}